drivers/network/dd/e1000/hardware.c

```diff
@@ -189,32 +190,83 @@ static BOOLEAN E1000ReadMdic(IN PE1000_ADAPTER Adapter, IN ULONG Address, USHORT
 }
 #endif
 
-
 static BOOLEAN E1000ReadEeprom(IN PE1000_ADAPTER Adapter, IN UCHAR Address, USHORT *Result)
 {
     ULONG Value;
     UINT n;
 
-    E1000WriteUlong(Adapter, E1000_REG_EERD, E1000_EERD_START | ((UINT)Address << E1000_EERD_ADDR_SHIFT));
+    if (Adapter->EERDAddrShift == 0xffff)
+    {
+        NDIS_DbgPrint(MIN_TRACE, ("E1000DetermineAndSaveEERDLayout not run or wasn't successfull!\n"));
+        return FALSE;
+    }
+    E1000WriteUlong(Adapter, E1000_REG_EERD, E1000_EERD_START | ((UINT)Address << Adapter->EERDAddrShift));
 
     for (n = 0; n < MAX_EEPROM_READ_ATTEMPTS; ++n)
     {
         NdisStallExecution(5);
 
         E1000ReadUlong(Adapter, E1000_REG_EERD, &Value);
 
-        if (Value & E1000_EERD_DONE)
+        if (Value & Adapter->EERDDoneBit)
             break;
     }
-    if (!(Value & E1000_EERD_DONE))
+    if (!(Value & Adapter->EERDDoneBit))
     {
-        NDIS_DbgPrint(MIN_TRACE, ("EEPROM Read incomplete\n"));
+        NDIS_DbgPrint(MIN_TRACE, ("EEPROM Read incomplete! %x \n", Value));
         return FALSE;
     }
     *Result = (USHORT)(Value >> E1000_EERD_DATA_SHIFT);
+    NDIS_DbgPrint(MIN_TRACE, ("Read: %02x %04x\n", Address, *Result));
     return TRUE;
 }
 
+/*
+ * 8254x and 8257x have different layouts of EERF
+ * Get it by trying to read the VID-field.
+ */
+static BOOLEAN E1000DetermineAndSaveEERDLayout(IN PE1000_ADAPTER Adapter)
+{
+    BOOLEAN Success;
+    USHORT Result = 0;
+
+    /* First, try the 8254x layout */
```

Comment:
This looks like guesswork to me, which is likely not backed by any official specification or production driver, is it?

The Linux e1000e driver simply has two different routines to read from the EEPROM, one for each product family: See https://elixir.bootlin.com/linux/v5.0/source/drivers/net/ethernet/intel/e1000e/nvm.c#L291 and https://elixir.bootlin.com/linux/v5.0/source/drivers/net/ethernet/intel/e1000e/ich8lan.c#L3301. This code also explains that the different structure results from the support of multiple flash banks in newer E1000 variants.
You already have access to an E1000_ADAPTER structure with a DeviceId field, so it should be simple to port that code structure to our e1000 driver.
I'm afraid we may otherwise hit subtle bugs in the future for unforeseen situations where that guesswork doesn't yield a valid result.